src/services/refactors/moveToFile.ts

```diff
@@ -0,0 +1,1114 @@
+
+    const imports = getNewFileImportsAndAddExportInOldFile(oldFile, newFile, usage.oldImportsNeededByNewFile, usage.newFileImportsFromOldFile, changes, checker, program, host, useEsModuleSyntax, quotePreference, importAdder);
+    const body = addExports(oldFile, toMove.all, usage.oldFileImportsFromNewFile, useEsModuleSyntax);
+    if (newFileExists) {
+        const sourceFile = program.getSourceFile(newFile);
+        if (sourceFile && sourceFile.statements.length > 0) {
+            changes.insertNodesAfter(sourceFile, sourceFile.statements[sourceFile.statements.length - 1], body);
+        }
+        if (sourceFile && imports.length > 0) {
+            insertImports(changes, sourceFile, imports, /*blankLineBetween*/ true, preferences);
+        }
+    }
+    if (importAdder) {
+        importAdder.writeFixes(changes);
+    }
+    if (imports.length && body.length) {
+        return [
+            ...prologueDirectives,
+            ...imports,
+            SyntaxKind.NewLineTrivia as const,
+            ...body
+        ];
+    }
+
+    return [
+        ...prologueDirectives,
+        ...imports,
+        ...body,
+    ];
+}
+
+function getNewFileImportsAndAddExportInOldFile(
+    oldFile: SourceFile,
+    newFile: string,
+    importsToCopy: Map<Symbol, boolean>,
+    newFileImportsFromOldFile: Set<Symbol>,
+    changes: textChanges.ChangeTracker,
+    checker: TypeChecker,
+    program: Program,
+    host: LanguageServiceHost,
+    useEsModuleSyntax: boolean,
+    quotePreference: QuotePreference,
+    importAdder?: codefix.ImportAdder,
+): readonly AnyImportOrRequireStatement[] {
+    const copiedOldImports: AnyImportOrRequireStatement[] = [];
+    if (importAdder) {
+        importsToCopy.forEach((isValidTypeOnlyUseSite, symbol) => {
+            try {
+                importAdder.addImportFromExportedSymbol(skipAlias(symbol, checker), isValidTypeOnlyUseSite);
+            }
+            catch {
+                for (const oldStatement of oldFile.statements) {
+                    forEachImportInStatement(oldStatement, i => {
+                        append(copiedOldImports, filterImport(i, factory.createStringLiteral(moduleSpecifierFromImport(i).text), name => importsToCopy.has(checker.getSymbolAtLocation(name)!)));
+                    });
+                }
+            }
+        });
+    }
+    else {
+        for (const oldStatement of oldFile.statements) {
+            forEachImportInStatement(oldStatement, i => {
+                const moduleSpecifier = moduleSpecifierFromImport(i);
+                const resolved = oldFile.resolvedModules?.get(moduleSpecifier.text, getModeForUsageLocation(oldFile, moduleSpecifier));
+                const fileName = resolved?.resolvedModule?.resolvedFileName;
+                if (fileName) {
+                    const sourceFile = program.getSourceFile(newFile);
+                    if (sourceFile) {

```

Comment:
Style nit:

Same as before, if `sourceFile` is `undefined`, something has gone horribly wrong and you can’t do anything useful to recover, so there’s no need to code defensively here. (Also, hoist this `program.getSourceFile` out so you’re not retrieving it repeatedly in this nested loop.)

Reply:
I've moved `program.getSourceFile` out but the `sourceFile` will be undefined if we are moving to a new file. So for recomputing the module specifier, if
- We move to an existing file, import adder will handle it
- If moving to a blank existing file, we will recompute it using `getModuleSpecifier()`
- If moving to a new file, it will get copied over as is because to use `getModuleSpecifier()` we need the `targetSourceFile` which is not available